File: py/hubspotconversations_sdk/features.py
# HubspotConversations SDK feature factory

from hubspotconversations_sdk.feature.base_feature import HubspotConversationsBaseFeature
from hubspotconversations_sdk.feature.debug_feature import HubspotConversationsDebugFeature
from hubspotconversations_sdk.feature.idempotency_feature import HubspotConversationsIdempotencyFeature
from hubspotconversations_sdk.feature.metrics_feature import HubspotConversationsMetricsFeature
from hubspotconversations_sdk.feature.paging_feature import HubspotConversationsPagingFeature
from hubspotconversations_sdk.feature.ratelimit_feature import HubspotConversationsRatelimitFeature
from hubspotconversations_sdk.feature.retry_feature import HubspotConversationsRetryFeature
from hubspotconversations_sdk.feature.test_feature import HubspotConversationsTestFeature
from hubspotconversations_sdk.feature.timeout_feature import HubspotConversationsTimeoutFeature


_FEATURES = {
    "base": lambda: HubspotConversationsBaseFeature(),
    "debug": lambda: HubspotConversationsDebugFeature(),
    "idempotency": lambda: HubspotConversationsIdempotencyFeature(),
    "metrics": lambda: HubspotConversationsMetricsFeature(),
    "paging": lambda: HubspotConversationsPagingFeature(),
    "ratelimit": lambda: HubspotConversationsRatelimitFeature(),
    "retry": lambda: HubspotConversationsRetryFeature(),
    "test": lambda: HubspotConversationsTestFeature(),
    "timeout": lambda: HubspotConversationsTimeoutFeature(),
}


def _make_feature(name):
    factory = _FEATURES.get(name)
    if factory is not None:
        return factory()
    return _FEATURES["base"]()


# True when this SDK was generated with the named feature class - the
# constructor's tolerance for extend-carried features reads this (an
# active name with no generated class must not become a BaseFeature
# stray when an extend instance carries it).
def _has_feature(name):
    return name in _FEATURES

File: php/features.php
<?php
declare(strict_types=1);

// HubspotConversations SDK feature factory

require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/feature/DebugFeature.php';
require_once __DIR__ . '/feature/IdempotencyFeature.php';
require_once __DIR__ . '/feature/MetricsFeature.php';
require_once __DIR__ . '/feature/PagingFeature.php';
require_once __DIR__ . '/feature/RatelimitFeature.php';
require_once __DIR__ . '/feature/RetryFeature.php';
require_once __DIR__ . '/feature/TestFeature.php';
require_once __DIR__ . '/feature/TimeoutFeature.php';


class HubspotConversationsFeatures
{
    public static function make_feature(string $name)
    {
        switch ($name) {
            case "base":
                return new HubspotConversationsBaseFeature();
            case "debug":
                return new HubspotConversationsDebugFeature();
            case "idempotency":
                return new HubspotConversationsIdempotencyFeature();
            case "metrics":
                return new HubspotConversationsMetricsFeature();
            case "paging":
                return new HubspotConversationsPagingFeature();
            case "ratelimit":
                return new HubspotConversationsRatelimitFeature();
            case "retry":
                return new HubspotConversationsRetryFeature();
            case "test":
                return new HubspotConversationsTestFeature();
            case "timeout":
                return new HubspotConversationsTimeoutFeature();
            default:
                return new HubspotConversationsBaseFeature();
        }
    }

    /**
     * Does a generated feature class back this name? False for a name only
     * an options extend instance can supply (the station adopt path) - the
     * constructor uses this to skip make_feature for such names instead of
     * adding a stray BaseFeature.
     */
    public static function has_feature(string $name): bool
    {
        switch ($name) {
            case "base":
            case "debug":
            case "idempotency":
            case "metrics":
            case "paging":
            case "ratelimit":
            case "retry":
            case "test":
            case "timeout":
                return true;
            default:
                return false;
        }
    }
}
